plugins/gpio/__init__.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+# vim: set encoding=utf-8 tabstop=4 softtabstop=4 shiftwidth=4 expandtab
+#########################################################################
+#  Copyright 2016 <Onkel Andy>                    <onkelandy@hotmail.com>
+#########################################################################
+#  This file is part of SmartHomeNG.   
+#
+#  SmartHomeNG is free software: you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation, either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  SmartHomeNG is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with SmartHomeNG. If not, see <http://www.gnu.org/licenses/>.
+#########################################################################
+
+import logging
+from lib.model.smartplugin import SmartPlugin
+import RPi.GPIO as GPIO
+import threading
+import time
+
+class Raspi_GPIO(SmartPlugin):
+    PLUGIN_VERSION = "1.0.0"
+    ALLOW_MULTIINSTANCE = False
+
+    def _wait(self,time_lapse):
+        time_start = time.time()
+        time_end = (time_start + time_lapse)
+     
+        while time_end > time.time():
+            pass
+
+    def __init__(self, sh, cycle=0, mode="board"):
+        self.logger = logging.getLogger(__name__)
+        self._sh = sh
+        self._items = []
+        self._cycle = int(cycle)
+        self._mode = mode.upper()
+        GPIO.setwarnings(False)
+        if self._mode == "BCM":
+            GPIO.setmode(GPIO.BCM)
+        else:
+            GPIO.setmode(GPIO.BOARD)    
+        self.logger.debug("GPIO: Mode set to {0}".format(self._mode))
+        self.alive = False
+        self._lock = threading.Lock() 
+        
+
+    def run(self):
+        self.logger.debug("run method called")
+        self.alive = True          
+        if self._cycle > 0:
+            self._sh.scheduler.add('Raspi_GPIO', self.get_sensors, prio=5, cycle=self._cycle)  
+        else:
+            self.logger.debug("GPIO: No Cycle defined, just running all the time.")
+            self.get_sensors() 
+            
+
+    def stop(self):
+        self.alive = False
+        try:
+            self._sh.scheduler.remove('Raspi_GPIO')
+        except:
+            self.logger.error("GPIO: Removing of scheduler failed: {}".format(sys.exc_info()))
+        self._wait(2)
+        GPIO.cleanup()
+        
+
+    def parse_item(self, item):
+        if self.has_iattr(item.conf, 'gpio_in'):
+            in_pin = int(item.conf['gpio_in'])
```

Comment:
Should we use `get_iattr_value()` instead of accessing the `conf` dict directly?

I still have a configuration dict in mind which can be used to specify which settings are used in this plugin and which type they are (e.g. `{ gpio_in : { desc : 'specifies the GPIO PIN for input', type : 'int', required : ..., ... }, gpio_out : ... }`). With this information we can
a) generate a better documentation and avoid double maintenance in code and README.md files and
b) automatically convert the values to the correct type (for checking existence use `has_iattr(name)`, for retrieving the value use `get_iattr(name)` - maybe the naming is not perfect and should be changed to `has_conf(name)` and `get_conf(name)`).